keps/sig-cli/2953-kustomize-plugin-graduation/README.md

```diff
@@ -0,0 +1,764 @@
+
+### Legacy plugin migration
+
+Although legacy plugins are still in alpha, the alpha has been around in standalone Kustomize for quite a while. Documentation on the migration process will be provided. See the [Rollout Plan](#rollout-plan) section for more detail on deprecation and removal of the existing alphas.
+
+The exec plugin conversion process should be relatively straightforward: copy the provider in-tree and update the configuration that referred to it to have a Provider stanza pointing to the new location. `argsFromFile` and `argsOneLiner` can be straightforwardly converted to the `args` subfield the KRM Functions style already supports.
+
+For authors who build plugins for distribution (as opposed to bespoke plugins where embedding within an end-user Kustomization is an appropriate solution), the recommendation will be to upgrade to the Catalog model. A Catalog guide will be added to the Kustomize website and will address the scenario of migrating a legacy style plugin.
+
+
+
+### User Stories (Optional)
+
+#### Story 1
+
+As an end user, I want to develop a simple plugin for my particular Kustomization.
+
+1. Write a script that adheres to the [KRM Functions Specification] and place it within the Kustomization root.
+2. Write a corresponding plugin config and reference it from the Kustomization's generator or transformer field as appropriate.
+3. Run `kustomize edit generate-catalog`
+3. Run `kustomize build --trusted-plugin-catalog=catalog.yaml`
+
+```bash
+# Possible directory structure
+.
+⊢ Kustomization.yaml
+⊢ input.yaml
+⊢ reorder.yaml
+⊢ plugin-catalog.yaml
+∟ plugins
+  ∟ sorter
+    ⊢ sorter.rb
+    ∟ sorter_test.rb
+```
+
+```yaml
+# Example Kustomization.yaml
+apiVersion: kustomize.config.k8s.io/v1beta1
+kind: Kustomization
+resources:
+- input.yaml
+transformers:
+- reorder.yaml
+```
+
+```yaml
+# Example reorder.yaml
+apiVersion: local-config.example.co/v1
+kind: CustomSorter
+provider:
+  exec: 
+  	path: plugins/sorter/sorter.rb
+spec:
+  order: NsGKVN
+```
+
+```yaml
+# Example catalog.yaml (auto-generated)
+
+apiVersion: kustomize.io/v1
+kind: Catalog
+metadata:
+  name: "local-plugins"
+spec:
+  modules:
+  - apiVersion: local-config.example.co/v1
+    kind: CustomSorter
+    provider:
+      exec:
+        - path: plugins/sorter/sorter.rb
+          sha256: [a hash]
+```
+
+
+#### Story 2
+
+As an end user, I want to use one or more plugins published by a third party I trust.
+
+1. Write the config objects for the plugins you want to use, and reference them from your Kustomization's generators or transformers field as appropriate.
+2. Run `kustomize build --trusted-plugin-catalog=https://catalog.kpt.dev/1.2.3.json`
+
+
+```bash
+# Possible directory structure
```

Comment:
How does it work when there are 2 `ConfigMap`s as the `functionConfig` for 2 different functions?

Reply:
We discussed this out of band and agreed that this pattern would not be supported. Typically in these cases, the input isn't really a ConfigMap (opaque key-value pairs that are not individually meaningful to the function), but rather a shadow type where there's a single layer of fields that all happen to have a string value. To work with Kustomize, plugins will need to use properly typed KRM, i.e. give each input type its own GVK. This is more consistent with Kustomize overall, and having properly specified types also opens the door to other KRM-driven features.